src/cunumeric/index/repeat.cu

```diff
@@ -0,0 +1,167 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "cunumeric/index/repeat.h"
+#include "cunumeric/index/repeat_template.inl"
+#include "cunumeric/cuda_help.h"
+
+#include <thrust/scan.h>
+#include <thrust/execution_policy.h>
+
+namespace cunumeric {
+
+using namespace Legion;
+
+template <typename Output, int DIM>
+static __global__ void __launch_bounds__(THREADS_PER_BLOCK, MIN_CTAS_PER_SM)
+  count_repeat_kernel(const size_t volume,
+                      Output out,
+                      const AccessorRO<int64_t, DIM> repeats,
+                      const Pitches<DIM - 1> pitches,
+                      const Point<DIM> origin,
+                      const size_t iters,
+                      Buffer<int64_t> offsets)
+{
+  int64_t value = 0;
+  for (size_t idx = 0; idx < iters; idx++) {
+    const size_t offset = (idx * gridDim.x + blockIdx.x) * blockDim.x + threadIdx.x;
+    if (offset < volume) {
+      auto point      = pitches.unflatten(offset, origin);
+      auto val        = repeats[point];
+      offsets[offset] = val;
+      SumReduction<int64_t>::fold<true>(value, val);
+    }
+  }
+  // Every thread in the thread block must participate in the exchange to get correct results
+  reduce_output(out, value);
+}
+
+template <typename VAL, int DIM>
+__global__ static void __launch_bounds__(THREADS_PER_BLOCK, MIN_CTAS_PER_SM)
+  repeat_kernel(Buffer<VAL> out,
+                const AccessorRO<VAL, DIM> in,
+                int64_t repeats,
+                const int32_t axis,
+                const Rect<DIM> rect,
+                const Pitches<DIM - 1> pitches,
+                const int volume)
+{
+  const size_t idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (idx >= volume) return;
+  auto p         = pitches.unflatten(idx, rect.lo);
+  size_t out_idx = idx * repeats;
+  for (size_t r = 0; r < repeats; r++) {
+    out[out_idx] = in[p];
+    ++out_idx;
+  }
+}
+
+template <typename VAL, int DIM>
+__global__ static void __launch_bounds__(THREADS_PER_BLOCK, MIN_CTAS_PER_SM)
+  repeat_kernel(Buffer<VAL> out,
+                const AccessorRO<VAL, DIM> in,
+                const AccessorRO<int64_t, DIM> repeats,
+                Buffer<int64_t> offsets,
+                const int32_t axis,
+                const Rect<DIM> rect,
+                const Pitches<DIM - 1> pitches,
+                const int volume)
+{
+  const size_t idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (idx >= volume) return;
+  auto p         = pitches.unflatten(idx, rect.lo);
+  size_t out_idx = offsets[idx];
+  for (size_t r = 0; r < repeats[p]; r++) {
```

Comment:
I see a similar issue as the previous one, though this one is even worse. that said, I don't have a better idea to write this code yet.

Reply:
I see 1 options for this code that will have better CUDA performance, but allocate more memory:
we can calculate offsets like:
`Buffer<int64_t> offsets = create_buffer<int64_t>(size, Memory::Kind::GPU_FB_MEM);`
here size is the size of an output array and then offsets will store information about which indx in the input array output index corresponds to
For example when `a = np.array([1, 2, 3, 4, 5])` and `repeats = np.array([1, 2, 1, 2, 1])`, offsets will have size 1 and be `[0,1,1,2,3,3,4] 
2This way we should be able to launch kernel with volume * repeats threads 
with no issues.

Reply:
However, I am not sure how to calculate offsets this way in performant way

Reply:
You can follow the pattern I had posted in the main discussion:

```
a = [10,11,12,13]
repeats = [1,4,2,3]
offsets = zeros((sum(repeats),)) = zeros((10,))
repeats = prefix_sum(repeats) # repeats is now [1,5,7,10]
offsets[repeats[:-1]] = 1 # offsets is now [0,1,0,0,0,1,0,1,0,0]
offsets = prefix_sum(offsets) # offsets is now [0,1,1,1,1,2,2,3,3,3]
output = a[offsets]
```

Reply:
And the corresponding OpenMP task could probably also be structured in the same way.

Reply:
Let's leave this for a follow-up PR.